packages/cozy-client/src/CozyClient.js

```diff
@@ -198,42 +198,51 @@ class CozyClient {
       event: TRIGGER_CREATION
     }
   ) {
-    if (!document._type) return document
-
-    const schema = this.schema.getDoctypeSchema(document._type)
-    if (!schema.cozyMetadata) return document
-
-    const enriched = {
-      ...document
-    }
-    const metadata = Object.entries(schema.cozyMetadata)
-
-    for (const [
-      metadataName,
-      { trigger, value, useCurrentDate = false }
-    ] of metadata) {
-      const shouldSetMetadata =
-        options.event === TRIGGER_CREATION ||
-        (options.event === TRIGGER_UPDATE && trigger === TRIGGER_UPDATE)
-      if (!shouldSetMetadata) continue
-
-      let newValue
-      const metadataPath = `cozyMetadata.${metadataName}`
-
-      const shouldAccumulateValues = Array.isArray(value)
-      if (shouldAccumulateValues) {
-        const prevValue = get(document, metadataPath, [])
-        newValue = uniqBy([...prevValue, ...value])
-      } else if (useCurrentDate) {
-        newValue = new Date().toISOString()
-      } else {
-        newValue = get(document, metadataPath, value)
+    if (this.metadata === undefined) return document
+    let doctypeVersion
+    if (document._type) {
+      const schema = this.schema.getDoctypeSchema(document._type)
+      doctypeVersion = get(schema, 'doctypeVersion')
+    }
+
+    const { slug, sourceAccount = null, version } = this.metadata
+
+    const now = new Date().toISOString()
+
+    let cozyMetadata = get(document, 'cozyMetadata', {})
+    if (options.event === TRIGGER_CREATION) {
+      cozyMetadata = {
+        ...cozyMetadata, // custom metadata that are set by the app
+        doctypeVersion,
+        createdByApp: slug,
+        sourceAccount,
+        createdAt: now,
+        createdByAppVersion: version,
+        updatedAt: now,
+        updatedByApps: [
+          {
+            date: now,
+            slug
+          }
+        ]
+      }
+    } else if (options.event === TRIGGER_UPDATE) {
+      cozyMetadata = {
+        ...cozyMetadata,
+        updatedAt: now,
+        updatedByApps: [
+          { date: now, slug },
```

Comment:
if "slug" is not mandatory, please avoid using it when not defined ?

Reply:
I think that if you provide `appMetadata`, `slug` should be mandatory (as multiple metadata rely on it). If you agree I can change the check at the beginning of `ensureCozyMetadata` by `if (get(this, 'appMetadata.slug') === undefined) return document`

Reply:
shouldn't we still set up `updatedAt` ans 'createdAt` anyways ? 

But otherwise, ok for me